firewood/src/merkle/stream.rs

```diff
@@ -88,197 +100,239 @@ impl<'a, S: ShaleStore<Node> + Send + Sync, T> Stream for MerkleKeyValueStream<'
+                            // node's children at [pos] or [pos + 1].
+                            if !path_to_key.is_empty() {
+                                // The next element in [path_to_key] will handle the child at [pos]
+                                // so we can start iterating at [pos + 1].
+                                let children_iter = get_children_iter(branch)
+                                    .filter(move |(_, child_pos)| child_pos > &pos);
+
+                                branch_iter_stack.push(BranchIterator {
+                                    key_nibbles: matched_key_nibbles.clone(),
+                                    children_iter: Box::new(children_iter),
+                                });
+                            } else {
+                                // Get the child at [pos], if any.
+                                let Some(child) = branch.children.get(pos as usize) else {
+                                    // This should never happen -- [pos] should never be OOB.
+                                    return Poll::Ready(Some(Err(api::Error::InternalError(
+                                        Box::new(api::Error::ChildNotFound),
+                                    ))));
+                                };
+
+                                let child = child
+                                    .map(|child_addr| merkle.get_node(child_addr))
+                                    .transpose()
+                                    .map_err(|e| api::Error::InternalError(Box::new(e)))?;
+
+                                let comparer = if child.is_none() {
+                                    // The child doesn't exist; we don't need to iterate over [pos].
+                                    |a: &u8, b: &u8| a > b
+                                } else {
+                                    // The child does exist; the first key to iterate over must be at [pos].
+                                    |a: &u8, b: &u8| a >= b
+                                };
+
+                                let children_iter = get_children_iter(branch)
+                                    .filter(move |(_, child_pos)| comparer(child_pos, &pos));
+
+                                branch_iter_stack.push(BranchIterator {
+                                    key_nibbles: matched_key_nibbles.clone(),
+                                    children_iter: Box::new(children_iter),
+                                });
+                            }
+
+                            matched_key_nibbles.push(pos);
+                        }
+                        NodeType::Extension(extension) => {
+                            if !path_to_key.is_empty() {
+                                // Add the extension node's path to the key nibbles.
+                                matched_key_nibbles.extend(extension.path.iter());
+                                continue;
+                            }
```

Comment:
Same comment as above. I think it would be easier to read the degenerate case as a separate branch 
```rust
NodeType::Extension(extension) if path_to_key.is_empty() => // ...
NodeType::Extension(extension) => // ...
```